src/main/telemetry/crsf.c

```diff
@@ -59,12 +59,72 @@
 
 #define CRSF_CYCLETIME_US                   100000 // 100ms, 10 Hz
 #define CRSF_DEVICEINFO_VERSION             0x01
-#define CRSF_DEVICEINFO_PARAMETER_COUNT     255
+#define CRSF_DEVICEINFO_PARAMETER_COUNT     0
+
+#define CRSF_MSP_BUFFER_SIZE 96
 
 static bool crsfTelemetryEnabled;
 static bool deviceInfoReplyPending;
 static uint8_t crsfFrame[CRSF_FRAME_SIZE_MAX];
 
+#if defined(USE_MSP_OVER_TELEMETRY)
+typedef struct mspFlatBuffer_s {
+    uint8_t buffer[CRSF_MSP_BUFFER_SIZE];
+    sbuf_t *head; // read-only
+    sbuf_t *tail; // write-only
+} mspFlatBuffer_t;
+
+static mspFlatBuffer_t mspFlatRxBuffer;
+static sbuf_t mspRxBufHead;
+static sbuf_t mspRxBufTail;
+
+void initCrsfMspBuffer(void)
+{
+    uint8_t *headStart = (uint8_t*)&mspFlatRxBuffer.buffer;
+    uint8_t *tailEnd = (uint8_t*)&mspFlatRxBuffer.buffer+CRSF_MSP_BUFFER_SIZE;
+    mspFlatRxBuffer.head = sbufInit(&mspRxBufHead, headStart, headStart);
+    mspFlatRxBuffer.tail = sbufInit(&mspRxBufTail, headStart, tailEnd);
+}
+
+bool bufferCrsfMspFrame(uint8_t *frameStart, int frameLength)
+{
+    bool frameBuffered = false;
+    if(sbufBytesRemaining(mspFlatRxBuffer.tail)>=frameLength) {
+        sbufWriteU8(mspFlatRxBuffer.tail, (uint8_t)frameLength);
+        sbufWriteData(mspFlatRxBuffer.tail, frameStart, frameLength);
+        sbufAdvance(mspFlatRxBuffer.tail, frameLength+1);
+        sbufInit(mspFlatRxBuffer.head, sbufPtr(mspFlatRxBuffer.head), sbufPtr(mspFlatRxBuffer.tail));
+        frameBuffered = true;
+    }
+    return frameBuffered;
+}
+
+bool handleCrsfMspBufferFrame(uint8_t payloadSize, mspResponseFnPtr responseFn)
+{
+    bool requestHandled = false;
+    while(sbufBytesRemaining(mspFlatRxBuffer.head)) {
+        const int mspFrameLength = (int)sbufReadU8(mspFlatRxBuffer.head);
+        if (mspFrameLength && mspFrameLength < CRSF_MSP_BUFFER_SIZE && sbufBytesRemaining(mspFlatRxBuffer.head)>=mspFrameLength) {
+            const int shiftLength = mspFrameLength+1;
```

Comment:
This is IMO quite overengineered ... 
It may be better to simple buffer + length of data in it. First packet is from address 0, after processing just memmove (memcpy does handle overlapping buffers) remaining data to start of buffer.

Reply:
```
if(bufferLen >=1 ) {
  const int mspFrameLength = buffer[0];
  (possibly check for len corruption)
  if (handleMspFrame(&buffer[1], mspFrameLength)) {
     requestHandled = sendMspReply(payloadSize, responseFn);
  }
  ATOMIC_BLOCK(...) {
    memmove(buffer, buffer + mspFrameLength + 1, bufferLen - (mspFrameLength + 1));
    bufferLen -= mspFrameLength + 1;
  }
}

Reply:
It's very possible that many frames could be appended to the buffer before deferred execution begins.  Some writes require up to 8 frames from the TX because the outgoing opentx buffer size only allows for 8 bytes of msp per crsf frame.

Reply:
not arguing that it's over-enginneered :)  I felt a bit like a mad scientist writing this.  Just looking for advice.

Reply:
My code example is a bit simplistic.

The idea is to use linear buffer that is filled in receiver (ISR), using `uncommittedLength` to track writing position. Each message is prepended by it's length (uint8_t), length is updated when committing message.
When packet is received (and possibly validated), frame size is written at `committedLength` position in buffer and `uncommittedLength` is copied into `committedLength`. 
On error `uncommittedLength` is reset back to `committedLength`, discarding received data.

When processing function is called, it will process all messages up to `committedLength`, in place/without copying. `committedLength` is advanced only on finished frame, so message and its length must be ready.
After processing all complete messages, rest of buffer (between `committedLength` and `uncommittedLength`) is memmoved to beginning of receive buffer, `committedLength` is set to zero and `uncommittedLength` is updated accordingly. This must be done atomically (it may be possible to sample/check `uncommittedLength` and move repeatedly, but it will be quite complicated - message commit/rollback etc).


This handling may be also implemented on serial level - use commited/uncommited head pointer + prepending of message length. Protocol decoder will just process escaping + frame boundary detection, receiver will use normal serial functions to access messages stored in port buffer ... It will greatly simplify implementation of some protocols ...

Reply:
I began taking your committed/uncommitted approach before I received the most recent commentary.  For the sake of time (and my need for better understanding at the serial level) to get this working at a slightly higher level and promote more serial level refinement in a later PR?  I'm afraid I may go down another over-engineering rabbit hole if attempting to go to far into uncharted territory.

Reply:
at what priority should the atomic block be called?

Reply:
(as a side-note, atomicity has not yet proven to be problematic in testing)

Reply:
Ad  serial-level frame handling - it's just an idea, the changeset would be quite large. Proving it with simpler, driver specific implementation (this PR) may be better.

NVIC_PRIO_SERIALUART1 should do (it should be highest serial  priority, but I won't complicate it now; softserial may need consideration too)

Ignoring race conditions is terrible approach - probability of race occurring is very low, it is unlikely that testing will catch it. And breaking frame queue may have catastrophic consequences.
You will end up with firmware that fails time to time, killing someone's quadcopter, but with probability too low to debug it.

Reply:
No problem.. I will update to add the extra protective measure.  The likelihood that a race condition would occur is extremely low, though, as people probably won't be looking down at their transmitter to tune it while they're flying, but who's to say someone wouldnt try :)

Reply:
The race condition should be handled for both writes and reads, correct?

Reply:
The race condition should be handled for both writes and reads, correct?

Reply:
It's safe to assume that `crsfDataReceive` is atomic - either it's called from ISR, or cooperative task scheduling will ensure atomicity.
Also it's safe to assume that integer variable reads are atomic.

Register caching without `volatile` is slightly complicated 
`ATOMIC_BLOCK` will place memory barriers, so value read from/written to variable shared with ISR will be correct.
Reading shared variable outside of `ATOMIC_BLOCK` will probably read it's content only once per function call - value is cached in register and reused.
In extreme cases `-fLTO` may merge multiple calls together, reading variable only once. Construct like `while(peekchar() < 0);` may be optimized as conditional jump + infinite loop if `peekchar()` is implemented without proper synchronization (`int available; int peekchar(void) { return available > 0; }`
In most cases function call is 'deep' enough to prevent this optimization.
Please read this article about `volatile`: https://github.com/Xilinx/linux-xlnx/blob/master/Documentation/volatile-considered-harmful.txt

Reply:
Understood.  I'll attempt to address the race condition first.

Reply:
`typedef struct mspBuffer_s {
    uint8_t bytes[CRSF_MSP_BUFFER_SIZE];
    int committedLength;
    int uncommittedLength;
} mspBuffer_t;

static mspBuffer_t mspRxBuffer;

void initCrsfMspBuffer(void)
{
    mspRxBuffer.committedLength = 0;
    mspRxBuffer.uncommittedLength = CRSF_MSP_BUFFER_SIZE;
}

bool bufferCrsfMspFrame(uint8_t *frameStart, int frameLength)
{
    bool frameBuffered = false;
    const int appendLength = CRSF_MSP_LENGTH_OFFSET + frameLength
    if (msbRxBuffer.uncommittedLength >= appendLength) {
        const uint8_t *tailPtr = (uint8_t*)&mspRxBuffer.bytes + mspRxBuffer.committedLength + 1
        *tailPtr++ = frameLength;
        memmove(tailPtr, frameStart, frameLength);
        mspRxBuffer.committedLength += appendLength;
        mspRxBuffer.uncommittedLength -= appendLength;
        frameBuffered = true;
    }
    return frameBuffered;
}

bool handleCrsfMspFrameBuffer(uint8_t payloadSize, mspResponseFnPtr responseFn)
{
    bool requestHandled = false;
    while (mspRxBuffer.bufferLen>0) {
        const int mspFrameLength = mspRxBuffer[0];
        const int shiftLength = mspFrameLength + 1;
        if(handleMspFrame(&mspRxBuffer.bytes[1], mspFrameLength)) {
            requestHandled = sendMspReply(payloadSize, responseFn);
        }
        ATOMIC_BLOCK(NVIC_PRIO_SERIALUART1) {
            memmove(mspRxBuffer.bytes, mspRxBuffer.bytes + shiftLength, CRSF_MSP_BUFFER_SIZE-shiftLength);
            mspRxBuffer.uncommittedLength += shiftLength;
            mspRxBuffer.committedLength -= shiftLength;
        }
    }
    return requestHandled;
}`

Reply:
```typedef struct mspBuffer_s {
    uint8_t bytes[CRSF_MSP_BUFFER_SIZE];
    int committedLength;
    int uncommittedLength;
} mspBuffer_t;

static mspBuffer_t mspRxBuffer;

void initCrsfMspBuffer(void)
{
    mspRxBuffer.committedLength = 0;
    mspRxBuffer.uncommittedLength = CRSF_MSP_BUFFER_SIZE;
}

bool bufferCrsfMspFrame(uint8_t *frameStart, int frameLength)
{
    bool frameBuffered = false;
    const int appendLength = CRSF_MSP_LENGTH_OFFSET + frameLength
    if (msbRxBuffer.uncommittedLength >= appendLength) {
        const uint8_t *tailPtr = (uint8_t*)&mspRxBuffer.bytes + mspRxBuffer.committedLength + 1
        *tailPtr++ = frameLength;
        memmove(tailPtr, frameStart, frameLength);
        mspRxBuffer.committedLength += appendLength;
        mspRxBuffer.uncommittedLength -= appendLength;
        frameBuffered = true;
    }
    return frameBuffered;
}

bool handleCrsfMspFrameBuffer(uint8_t payloadSize, mspResponseFnPtr responseFn)
{
    bool requestHandled = false;
    while (mspRxBuffer.bufferLen>0) {
        const int mspFrameLength = mspRxBuffer[0];
        const int shiftLength = mspFrameLength + 1;
        if(handleMspFrame(&mspRxBuffer.bytes[1], mspFrameLength)) {
            requestHandled = sendMspReply(payloadSize, responseFn);
        }
        ATOMIC_BLOCK(NVIC_PRIO_SERIALUART1) {
            memmove(mspRxBuffer.bytes, mspRxBuffer.bytes + shiftLength, CRSF_MSP_BUFFER_SIZE-shiftLength);
            mspRxBuffer.uncommittedLength += shiftLength;
            mspRxBuffer.committedLength -= shiftLength;
        }
    }
    return requestHandled;
}```

Reply:
```typedef struct mspBuffer_s {
    uint8_t bytes[CRSF_MSP_BUFFER_SIZE];
    int committedLength;
    int uncommittedLength;
} mspBuffer_t;

static mspBuffer_t mspRxBuffer;

void initCrsfMspBuffer(void)
{
    mspRxBuffer.committedLength = 0;
    mspRxBuffer.uncommittedLength = CRSF_MSP_BUFFER_SIZE;
}

bool bufferCrsfMspFrame(uint8_t *frameStart, int frameLength)
{
    bool frameBuffered = false;
    const int appendLength = CRSF_MSP_LENGTH_OFFSET + frameLength
    if (msbRxBuffer.uncommittedLength >= appendLength) {
        const uint8_t *tailPtr = (uint8_t*)&mspRxBuffer.bytes + mspRxBuffer.committedLength + 1
        *tailPtr++ = frameLength;
        memmove(tailPtr, frameStart, frameLength);
        mspRxBuffer.committedLength += appendLength;
        mspRxBuffer.uncommittedLength -= appendLength;
        frameBuffered = true;
    }
    return frameBuffered;
}

bool handleCrsfMspFrameBuffer(uint8_t payloadSize, mspResponseFnPtr responseFn)
{
    bool requestHandled = false;
    while (mspRxBuffer.bufferLen>0) {
        const int mspFrameLength = mspRxBuffer[0];
        const int shiftLength = mspFrameLength + 1;
        if(handleMspFrame(&mspRxBuffer.bytes[1], mspFrameLength)) {
            requestHandled = sendMspReply(payloadSize, responseFn);
        }
        ATOMIC_BLOCK(NVIC_PRIO_SERIALUART1) {
            memmove(mspRxBuffer.bytes, mspRxBuffer.bytes + shiftLength, CRSF_MSP_BUFFER_SIZE-shiftLength);
            mspRxBuffer.uncommittedLength += shiftLength;
            mspRxBuffer.committedLength -= shiftLength;
        }
    }
    return requestHandled;
}
```

Reply:
```
typedef struct mspBuffer_s {
    uint8_t bytes[CRSF_MSP_BUFFER_SIZE];
    int committedLength;
    int uncommittedLength;
} mspBuffer_t;

static mspBuffer_t mspRxBuffer;

void initCrsfMspBuffer(void)
{
    mspRxBuffer.committedLength = 0;
    mspRxBuffer.uncommittedLength = CRSF_MSP_BUFFER_SIZE;
}

bool bufferCrsfMspFrame(uint8_t *frameStart, int frameLength)
{
    bool frameBuffered = false;
    const int appendLength = CRSF_MSP_LENGTH_OFFSET + frameLength
    if (msbRxBuffer.uncommittedLength >= appendLength) {
        const uint8_t *tailPtr = (uint8_t*)&mspRxBuffer.bytes + mspRxBuffer.committedLength + 1
        *tailPtr++ = frameLength;
        memmove(tailPtr, frameStart, frameLength);
        mspRxBuffer.committedLength += appendLength;
        mspRxBuffer.uncommittedLength -= appendLength;
        frameBuffered = true;
    }
    return frameBuffered;
}

bool handleCrsfMspFrameBuffer(uint8_t payloadSize, mspResponseFnPtr responseFn)
{
    bool requestHandled = false;
    while (mspRxBuffer.bufferLen>0) {
        const int mspFrameLength = mspRxBuffer[0];
        const int shiftLength = mspFrameLength + 1;
        if(handleMspFrame(&mspRxBuffer.bytes[1], mspFrameLength)) {
            requestHandled = sendMspReply(payloadSize, responseFn);
        }
        ATOMIC_BLOCK(NVIC_PRIO_SERIALUART1) {
            memmove(mspRxBuffer.bytes, mspRxBuffer.bytes + shiftLength, CRSF_MSP_BUFFER_SIZE-shiftLength);
            mspRxBuffer.uncommittedLength += shiftLength;
            mspRxBuffer.committedLength -= shiftLength;
        }
    }
    return requestHandled;
}
```

Reply:
haven't compiled this yet -- just reaching out for feedback before I go much further

Reply:
compiled and tested successfully

Reply:
Hmm.. I'll go ahead and push for review by the unit tests are choking on the call of ATOMIC_BLOCK.  Expect failures in the test step.

Reply:
../main/build/atomic.h:26:4: error: use of undeclared identifier '__ASM'

Reply:
I was a bit ahead with my algorithm - I was expecting that received bytes are stored into correct buffer space directly, without intermediate CRSF buffer. That is unnecessarily complicated for now.

When adding whole packets atomically (called from receive handler), it is simpler - just keep used buffer length, `memcpy`ing received packets to end of buffer (after prepending length).

When handling frames, process all available frames in place, no copy is necessary.
There is no need to move rest of buffer - just atomically check if all data are processed and  reset length.

```
bool bufferCrsfMspFrame(uint8_t *frameStart, int frameLength)
{
    if (mspRxBuffer.len + frameLength + sizeof(uint8_t) > CRSF_MSP_BUFFER_SIZE) {
        return false;  // bail out
    }
    uint8_t *p = mspRxBuffer.bytes + mspRxBuffer.len;
    *p++ = frameLength;
    memcpy(p, frameStart, frameLength);
    mspRxBuffer.len += frameLength + 1;
    return true;
}

bool handleCrsfMspFrameBuffer(uint8_t payloadSize, mspResponseFnPtr responseFn)
{
    bool requestHandled = false;
    if(!mspRxBuffer.len) {
       // fast path on empty buffer
       return requestHandled;
    }
    int pos = 0;
    while(true) {
        const int mspFrameLength = mspRxBuffer.bytes[pos];
        if(handleMspFrame(&mspRxBuffer.bytes[pos + 1], mspFrameLength)) {
            requestHandled |= sendMspReply(payloadSize, responseFn);
        }
        pos += mspFrameLength + 1;
        ATOMIC_BLOCK(NVIC_PRIO_SERIALUART1) {
             if(pos =>= mspRxBuffer.len) { // handle pos overflow gracefully
                 // all frames processed
                 mspRxBuffer.len = 0;   // flush buffer
                 return requestHandled;
             }
             // something left in buffer, continue processing frames
        }
    }
}

Reply:
Problem with unittests is strange - ATOMIC_BLOCK should expand to normal block in unittests ... I'll fix it ASAP
```
#define ATOMIC_BLOCK(x) for(__i = 1; __i; __i = 0)
```
(copy macro magic to get unique identifier instead of `__i`

Reply:
Wouldn't shifting the buffer prevent overflows if additional frames come in?  Or do you think the frequency of the buffer consumer is high enough that it would not create the potential for overflow?

Reply:
Or is it possible that a frame could come in and be nuked by the buffer reset before being processed?

Reply:
I assume the atomic_block will prevent this from happening.

Reply:
Hmm.. overflow I think.. hung my debugger.

Reply:
Frame processing is fast when scheduler gets to it (nothing can preempt it), and receive buffer size should allow some reserve. Moving packets inside loop will take some CPU time and gain will be very small (handleMspFrame in most cases only copies data into correct position, it is very fast).
Occasional overflow is OK, if receive overflow occurs too often, scheduling frequency has to be increased.

Check and reset of  `mspRxBuffer.len` is atomic, either new packet processed or buffer position is reset.